### PR TITLE
Fix faulty return value conversion for socket io

### PR DESCRIPTION
**Pull request by v-atamanenko**

### CODE REVIEW

File: newlib/libc/sys/vita/syscalls.c [hunk 1]
```diff
@@ -42,8 +43,13 @@ _write_r(struct _reent * reent, int fd, const void *buf, size_t nbytes)
 		ret = sceIoWrite(fdmap->sce_uid, buf, nbytes);
 		break;
 	case VITA_DESCRIPTOR_SOCKET:
-		if (__vita_glue_socket_send)
-			ret = __vita_glue_socket_send(fdmap->sce_uid, buf, nbytes, 0);
+		ret = sceNetSend(fdmap->sce_uid, buf, nbytes, 0);
+		if (ret < 0) {
+			if (ret != -1 )
```
Comment (isage) [inline newlib/libc/sys/vita/syscalls.c]: Is that needed?

File: newlib/libc/sys/vita/syscalls.c [hunk 2]
```diff
@@ -265,8 +271,13 @@ _read_r(struct _reent *reent, int fd, void *ptr, size_t len)
 		ret = sceIoRead(fdmap->sce_uid, ptr, len);
 		break;
 	case VITA_DESCRIPTOR_SOCKET:
-		if (__vita_glue_socket_recv)
-			ret = __vita_glue_socket_recv(fdmap->sce_uid, ptr, len, 0);
+		ret = sceNetRecv(fdmap->sce_uid, ptr, len, 0);
+		if (ret < 0) {
+			if (ret != -1 )
```
Comment (isage) [inline newlib/libc/sys/vita/syscalls.c]: Same

**Comment by isage** [comment]
Remove glue funcs too if they aren't used anymore?

**Comment by isage** [comment]
Looks good.

**Comment by v-atamanenko** [comment]
tested, works well

**Comment by v-atamanenko** [comment]
To the merging person: please add @isage in Co-authored-by: in squash commit

**Comment by frangarcj** [comment]
@d3m3vilurr you did the original code, can you take a look?

I see some lw and fd drop calls that are new

**Comment by v-atamanenko** [comment]
these are not really new, just moved from either now-removed nested vitaglue function, or copied from the bottom of the function (since we need to return earlier). the behavior isn't changed in this sense.

File: newlib/libc/sys/vita/socket.c [hunk 3]
```diff
@@ -467,22 +467,4 @@ int	socket(int domain, int type, int protocol)
 	__vita_fdmap[s]->type = VITA_DESCRIPTOR_SOCKET;
 	return s;
 }
-
-int __vita_glue_socket_close(SceUID scefd)
```
Comment (d3m3vilurr) [inline newlib/libc/sys/vita/socket.c:471]: personally i want to keep this week reference.

**Review by d3m3vilurr — COMMENTED** [review]
i cannot find any changes of this PR.
it seems just remove week reference isn't it?

could you explain actual changing part?

**Comment by v-atamanenko** [comment]
@d3m3vilurr previously, sceErrNo got converted in vitaglue functions and returned from there back to syscalls. This caused the behavior I experienced: when trying to read() something bigger then the buffer, read should return -1 and EAGAIN in errno. But in this case it returned 11 and errno 0, as if there was 11 bytes read. This PR fixes read, write, and close which are all prone to this error.

You can read even more detailed story in the henkaku discord, we discussed the issue there with isage.
https://discord.com/channels/439481392548675594/465960127963463680/897881471434522734

On the side note, what’s the point of a weak ref here?

**Pull request merged**